### DISCUSSION TITLE
Array of objects doesnt update state

### DISCUSSION THREAD

**Discussion by yixe** [pointfreeco/swift-composable-architecture#423] [Q&A]
Hi guys, I have a simple state:

```swift
struct AppState: Equatable {
  var collection: [Single] = []
 }

struct Single: Equatable {
  var name: String
  var value: Int
 }
```
Whenever I send an action to update the value on an Item in the Array the value updates but isn't reflected in the view. I've tried using ForEachStore etc but nothing seems to work

**Comment by lukeredpath** [comment]
You'll need to share a bit more code to help us debug what's happening but assuming you have some kind of child reducer that operates on `Single` value types, remember that the child reducer is going to be mutating a copy of each element in the array. Swift arrays are not reference types, and you can demonstrate the problem more generally:

```swift
var collection: [Single] = [
  .init(name: "one", value: 1),
  .init(name: "two", value: 2)
]

var first = collection[0]
first.name = "one modified"

print(collection[0]) // => Single(name: "one", value: 1)
```

**Reply by yixe** [reply]
Thanks for your help @lukeredpath, I'm confused as to why the state isnt changing in the view but .debug() on the reducer seems to show the value changing. 
In my reducer I'm simply modifying the value like so:
```swift
...
case .changeValue(let index):
state.collection[index].value = state.collection[index].value + 1
return .none
...
```

**Reply by lukeredpath** [reply]
Swift values are copy on write, so when you do:

```swift
state.collection[index].value = ANYTHING
```

You aren't mutating `state`, but instead a copy of `collection[index]` (your `Single` value). Because you aren't keeping a reference to this copy anywhere, that mutated value is lost. 

If all you're trying to do is increment the value, you could do this:

```swift
case .changeValue(let index):
  var single = state.collection[index]
  single.value += 1
  state.collection[index] = single
```

You could refactor this into a mutating helper method on your state:

```swift
struct AppState {
  mutating func incrementSingle(at index: Int) {
    var single = collection[index]
    single.value += 1
    collection[index] = single
  }
}

// in your reducer
case .changeValue(let index):
  state.incrementSingle(at: index)
```

**Reply by RabugenTom** [reply]
@lukeredpath Swift values are not all copy on write though. Arrays are however, but this doesn't affect the value type semantic, as this is only an implementation detail. When you execute
```swift
state.collection[index].value = state.collection[index].value + 1 // or += 1
```
you're accessing a specific setter of `state` on the lhs of `=`, and the whole state is modified.
This can be verified in a playground, or even by the compiler if you define your state as a constant `let state = ...`.

@yixe I would try to make it work without the array and the `ForEach` to begin with.  Define a single `Single` state in your app state and try to make the UI work for this. The array thing adds a little more complexity, so I would simplify the problem until I get something working, and then elaborate incrementally from there.

**Reply by lukeredpath** [reply]
@RabugenTom in my orthopaedic shoes, I stand corrected.

Then it's not immediately clear to me from the code that's been posted what is going wrong, I feel like we need to see a more complete example from OP to help.

**Reply by yixe** [reply]
@RabugenTom @lukeredpath I've been trying to debug this for a few days now, Whats bizarre to me is when you log the output of state, the value in fact changes.
```swift
case .changeValue(let index):
  state.collection[index].value +=  1
  dump(state.collection[index])  <-- outputs a Single with the +1 change
  return .none
```
but the view is stuck on the old value, @lukeredpath this seems to happen even when you make Single a class
```swift 
Text("\(viewStore.collection[index].value)")
```

**Reply by yixe** [reply]
@RabugenTom Following your recommendation, It seems I might have figured out the problem
If I add my own Equatable conformance like so or make Single a class the issue occurs. Quite confusing.

```swift
  static func == (lhs: Single, rhs: Single) -> Bool {
    lhs.name == rhs.name && lhs.value == rhs.value
  }

```

**Reply by RabugenTom** [reply]
I would absolutely recommend to use value types for your state. This is only a logic-less container for your data so you shouldn't have to use a class if you're are not forced to by a very specific reason. The same for automatic `Equatable` and `Hashable` conformances.
Your `Equatable` implementation is most likely similar to the one the compiler would generate so maybe this is only the expression of a deeper issue somewhere else.
Good luck!

**Reply by konrain** [reply]
Thanks for your help
